pkg/playback/http_playback_test.go

```diff
@@ -0,0 +1,275 @@
+	mockResponse1.Headers.Add("testHeader", "testHeaderValue")
+	mockResponse1.Body = []byte("body 1")
+
+	mockResponse2 := httpResponse{}
+	mockResponse2.StatusCode = 402
+	mockResponse2.Headers = make(http.Header)
+	mockResponse2.Body = []byte("body 2")
+	fixtureResponses := []httpResponse{mockResponse1, mockResponse2}
+
+	ts := startMockServer(fixtureResponses)
+	defer ts.Close()
+	remoteURL = ts.URL
+
+	// -- Setup Playback server
+	addressString := "localhost:13111"
+	cassetteFilepath := "test_record_replay_single_run.yaml"
+
+	// for now, write cassettes to this directory. Ideally we have a test output folder
+	cassetteDirectory, err := filepath.Abs("")
+	assert.NoError(t, err)
+
+	webhookURL := defaultLocalWebhookAddress // not used in this test
+	httpWrapper, err := NewServer(remoteURL, webhookURL, cassetteDirectory, Record, cassetteFilepath)
+	assert.NoError(t, err)
+
+	server := httpWrapper.InitializeServer(addressString)
+	go func() {
+		server.ListenAndServe()
+	}()
+	defer server.Shutdown(context.TODO())
+
+	fullAddressString := "http://" + addressString
+	// --- Start interacting in RECORD MODE
+
+	// Send it 2 requests
+	res1, err := http.Get(fullAddressString + "/r1")
+	assert.NoError(t, err)
+	assert.Equal(t, mockResponse1.StatusCode, res1.StatusCode)
+
+	res2, err := http.Get(fullAddressString + "/r2")
+	assert.NoError(t, err)
+	assert.Equal(t, mockResponse2.StatusCode, res2.StatusCode)
+
+	// Tell server to save recording
+	resp, err := http.Get(fullAddressString + "/playback/cassette/eject")
+	assert.NoError(t, err)
+	defer resp.Body.Close()
+	assert.Equal(t, 200, resp.StatusCode)
+
+	// --- END RECORD MODE
+
+	// --- Start interacting in REPLAY MODE
+	resp, err = http.Get(fullAddressString + "/playback/mode/replay")
+	assert.NoError(t, err)
+	defer resp.Body.Close()
+	assert.Equal(t, 200, resp.StatusCode)
+
+	resp, err = http.Get(fullAddressString + "/playback/cassette/load?filepath=" + cassetteFilepath)
+	assert.NoError(t, err)
+	defer resp.Body.Close()
+	assert.Equal(t, 200, resp.StatusCode)
+
+	replay1, err := http.Get(fullAddressString + "/r1")
+	assert.NoError(t, err)
+	replay2, err := http.Get(fullAddressString + "/r2")
+	assert.NoError(t, err)
+
+	// Assert that the original responses and replayed responses are the same
+	check(t, assertHTTPResponsesAreEqual(t, res1, replay1))
+	check(t, assertHTTPResponsesAreEqual(t, res2, replay2))
+
+	// Also sanity check that the mock server is responding with the expected responses
+	assert.Equal(t, "testHeaderValue", res1.Header.Get("testHeader"))
+	bodyBytes1, err := ioutil.ReadAll(res1.Body)
+	assert.NoError(t, err)
+	assert.Equal(t, mockResponse1.Body, bodyBytes1)
+
+	bodyBytes2, err := ioutil.ReadAll(res2.Body)
+	assert.NoError(t, err)
+	assert.Equal(t, mockResponse2.Body, bodyBytes2)
+
+	// Cleanup file
+	_, err = os.Stat(cassetteFilepath)
+	if err == nil {
+		os.Remove(cassetteFilepath)
+	}
+
+	// --- END REPLAY MODE
+}
+
+// // TODO: Test auto mode on the full server
```

Comment:
we can track the TODOs in a Jira maybe, let's remove the comments.

Reply:
Done - I kept the surviving TODOs but linked them to their Jira tickets so that's its easier for future people to identify where the changes should be made. See b698ea234bb8be7e5d9d2e99d3ffca8356541223.